dali/pipeline/graph_descr.cc

```diff
@@ -71,358 +73,444 @@ void CheckOpConstraints(const OpSpec &spec) {
+// 2. Remove tensors it produces
+// 3. Remove us from consumer lists of parent tensors
+// 4. Remove the OpNode with edges from parent Ops
+// 6. Correct the partitions as we changed the ids while removing OpNode
+void OpGraph::RemoveOp(OpNodeId id) {
+  OpNode &target = this->Node(id);
+
+  // If the node has any children, we cannot remove it
+  DALI_ENFORCE(target.children.empty(), "Node '" + target.spec.name() +
+      "' has " + std::to_string(target.children.size()) +
+      ". Cannot remove");
+  for (auto t : target.children_tensors) {
+    DALI_ENFORCE(tensor_nodes_[t].consumer_edges.empty(), "Node '" + target.spec.name() +
+      "' produces a tensor that has " + std::to_string(tensor_nodes_[t].consumer_edges.size()) +
+      " consumers. Cannot remove");
+  }
 
-    for (size_t i = idx; i < support_nodes_.size(); ++i) {
-      OpNode &support_node = this->support_node(i);
-      id_to_node_map_[support_node.id].second = i;
+  // Remove all tensors produced by this node and invalidate list of children tensors
+  for (auto t : target.children_tensors) {
+    RemoveTensorNode(t);
+  }
+  target.children_tensors.clear();
+
+  // In case we consume this tensor more than once, we try to remove all occurences
+  for (auto t : target.parent_tensors) {
+    auto &sibling_consumers = tensor_nodes_[t].consumer_edges;
+    for (int i = 0; i < sibling_consumers.size(); i++) {
+      if (sibling_consumers[i].node == id) {
+        RemoveVectorElement(sibling_consumers, i);
+      }
     }
   }
+
+  RemoveOpNode(id);
+  Repartition();
 }
 
-OpNode& OpGraph::node(NodeID id) {
-  DALI_ENFORCE_VALID_INDEX(id, id_to_node_map_.size());
-  auto idx_pair = id_to_node_map_[id];
-
-  switch (idx_pair.first) {
-  case DALI_CPU:
-    return cpu_nodes_[idx_pair.second];
-    break;
-  case DALI_GPU:
-    return gpu_nodes_[idx_pair.second];
-    break;
-  case DALI_MIXED:
-    return mixed_nodes_[idx_pair.second];
-    break;
-  case DALI_SUPPORT:
-    return support_nodes_[idx_pair.second];
-    break;
-  default:
-    DALI_FAIL("Internal error. Invalid node type index.");
+void OpGraph::Repartition() {
+  for (auto & p : node_partitions_) {
+    p.clear();
+  }
+  for (auto &node : op_nodes_) {
+    auto new_partition_id = NumOp(node.op_type);
+    node.partition_index = new_partition_id;
+    node_partitions_[static_cast<int>(node.op_type)].push_back(node.id);
   }
 }
 
-OpNode& OpGraph::node(const std::string& name) {
-  // Search cpu nodes
-  for (auto& node : cpu_nodes_) {
+// TODO(klecki): get rid of string indexing
+OpNode& OpGraph::Node(const std::string& name) {
+  for (auto &node : op_nodes_) {
     if (node.instance_name == name) {
       return node;
     }
   }
-  // Search gpu nodes
-  for (auto& node : gpu_nodes_) {
-    if (node.instance_name == name) {
-      return node;
+  DALI_FAIL("Operator node with name " + name + " not found.");
+}
+
+namespace {
+  std::ofstream& PrintTo(std::ofstream &ofs, const OpNode& node, bool show_ids) {
```

Comment:
Please add one line of documentation. Spending 5 sec on reading one sentence is better, than spending 2 min figuring it out from the code ;)